plugins/home/src/components/VisitListener.tsx

```diff
@@ -0,0 +1,163 @@
+ *   // returns "component:default/playback-order"
+ */
+export const getToEntityRef =
+  ({
+    rootPath = 'catalog',
+    stringifyEntityRefImpl = stringifyEntityRef,
+  } = {}) =>
+  ({ pathname }: { pathname: string }): string | undefined => {
+    const regex = new RegExp(
+      `^\/${rootPath}\/(?<namespace>[^\/]+)\/(?<kind>[^\/]+)\/(?<name>[^\/]+)`,
+    );
+    const result = regex.exec(pathname);
+    if (!result || !result?.groups) return undefined;
+    const entity = {
+      namespace: result.groups.namespace,
+      kind: result.groups.kind,
+      name: result.groups.name,
+    };
+    return stringifyEntityRefImpl(entity);
+  };
+
+/**
+ * @public
+ * This function returns an implementation of visitName which is responsible
+ * for receiving a pathname and returning a string (name). The default
+ * implementation ignores the pathname and uses the document.title .
+ */
+export const getVisitName = (document: Document) => () => document.title;
+
+/**
+ * @public
+ * Component responsible for listening to location changes and calling
+ * the visitsApi to save visits.
+ */
+export const VisitListener = ({
+  children,
+  toEntityRef,
+  visitName,
+}: {
+  children?: React.ReactNode;
+  toEntityRef?: ({ pathname }: { pathname: string }) => string | undefined;
+  visitName?: ({ pathname }: { pathname: string }) => string;
+}): JSX.Element => {
+  const [doNotTrack, setDoNotTrack] = useState<boolean>(
+    defaultVisitListenerContext.doNotTrack,
+  );
+  const visitsApi = useApi(visitsApiRef);
+  const { pathname } = useLocation();
+  const toEntityRefImpl = toEntityRef ?? getToEntityRef();
+  const visitNameImpl = visitName ?? getVisitName(document);
+  useEffect(() => {
+    // Wait for the browser to finish with paint with the assumption react
+    // has finished with dom reconciliation and the doNotTrack state update.
+    const requestId = requestAnimationFrame(() => {
+      if (!doNotTrack)
+        visitsApi.saveVisit({
+          visit: {
+            name: visitNameImpl({ pathname }),
+            pathname,
+            entityRef: toEntityRefImpl({ pathname }),
+          },
+        });
+    });
+    return () => cancelAnimationFrame(requestId);
+  }, [doNotTrack, visitsApi, pathname, toEntityRefImpl, visitNameImpl]);
+
+  return (
+    <VisitListenerContext.Provider value={{ doNotTrack, setDoNotTrack }}>
+      {children}
+    </VisitListenerContext.Provider>
+  );
+};
+
+/**
+ * @public
+ * Hook used to access visit listener context. Is able to control if tracking
+ * should be disabled.
+ */
+export const useVisitListener = () => {
+  const value = useContext(VisitListenerContext);
+
+  if (value === undefined)
+    throw new Error(
+      'useVisitListener found an undefined context, <VisitListener/> could be missing',
+    );
+
+  return value;
+};
+
+/**
+ * @public
+ * Use this component to warn VisitListener to disable tracking.
+ */
+export const DoNotTrack = ({
```

Comment:
Is the reason for not tracking to save storage space? If not, I wonder if the unwanted records can be filtered out in the visits query itself.

Reply:
Makes sense, thanks for explaining.